rpc/src/builder.rs

```diff
@@ -0,0 +1,505 @@
+        Self::default().with_http(http)
+    }
+
+    /// Creates a new config with only ws set
+    pub fn set_ws(ws: impl Into<RpcModuleSelection>) -> Self {
+        Self::default().with_ws(ws)
+    }
+
+    /// Creates a new config with only ipc set
+    pub fn set_ipc(ipc: impl Into<RpcModuleSelection>) -> Self {
+        Self::default().with_ipc(ipc)
+    }
+
+    /// Sets the [RpcModuleSelection] for the http transport.
+    pub fn with_http(mut self, http: impl Into<RpcModuleSelection>) -> Self {
+        self.http = Some(http.into());
+        self
+    }
+
+    /// Sets the [RpcModuleSelection] for the ws transport.
+    pub fn with_ws(mut self, ws: impl Into<RpcModuleSelection>) -> Self {
+        self.ws = Some(ws.into());
+        self
+    }
+
+    /// Sets the [RpcModuleSelection] for the http transport.
+    pub fn with_ipc(mut self, ipc: impl Into<RpcModuleSelection>) -> Self {
+        self.ipc = Some(ipc.into());
+        self
+    }
+
+    /// Returns true if no transports are configured
+    pub fn is_empty(&self) -> bool {
+        self.http.is_none() && self.ws.is_none() && self.ipc.is_none()
+    }
+
+    /// Returns the [RpcModuleSelection] for the http transport
+    pub fn http(&self) -> Option<&RpcModuleSelection> {
+        self.http.as_ref()
+    }
+
+    /// Returns the [RpcModuleSelection] for the ws transport
+    pub fn ws(&self) -> Option<&RpcModuleSelection> {
+        self.ws.as_ref()
+    }
+
+    /// Returns the [RpcModuleSelection] for the ipc transport
+    pub fn ipc(&self) -> Option<&RpcModuleSelection> {
+        self.ipc.as_ref()
+    }
+
+    /// Ensures that both http and ws are configured and that they are configured to use the same
+    /// port.j
+    pub(crate) fn ensure_ws_http_identical(&self) -> Result<(), WsHttpSamePortError> {
+        if RpcModuleSelection::are_identical(self.http.as_ref(), self.ws.as_ref()) {
+            Ok(())
+        } else {
+            let http_modules = self
+                .http
+                .clone()
+                .map(RpcModuleSelection::into_selection)
+                .unwrap_or_default();
+            let ws_modules = self
+                .ws
+                .clone()
+                .map(RpcModuleSelection::into_selection)
+                .unwrap_or_default();
+            Err(WsHttpSamePortError::ConflictingModules {
+                http_modules,
+                ws_modules,
+            })
+        }
+    }
+}
+
+/// Describes the modules that should be installed.
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+pub enum RpcModuleSelection {
+    /// Use _all_ available modules.
+    All,
+    /// The default modules `discv5`, `history`, `web3`
+    #[default]
+    Standard,
+    /// Only use the configured modules.
+    Selection(Vec<PortalRpcModule>),
+}
+
+impl RpcModuleSelection {
+    /// The standard modules to instantiate by default `discv5`, `history`, `web3`
+    pub const STANDARD_MODULES: [PortalRpcModule; 3] = [
```

Comment:
Thinking further about these "default" modules... don't we **need** to instantiate the beacon network by default? Otherwise we have no way to validate data returned from the history network (post-merge)

Reply:
Hmm, yeah, I think launching Beacon API should be the default if we use to for history validation.